tinygrad/shape/shapetracker.py

```diff
@@ -2,99 +2,109 @@
 from enum import Enum, auto
 import functools
-from typing import Dict, Tuple, Union, List, Optional, Callable, cast
+from typing import Dict, Tuple, Union, List, Optional, Callable, NamedTuple
 from tinygrad.helpers import prod, DEBUG
 from tinygrad.shape.symbolic import Variable, MulNode, NumNode, Node, SumNode
 
 # these ops live here
 class MovementOps(Enum): RESHAPE = auto(); PERMUTE = auto(); EXPAND = auto(); PAD = auto(); SHRINK = auto(); STRIDE = auto() # noqa: E702
 
 @functools.lru_cache(maxsize=None)
-def to_shape_strides(shape:Tuple[int, ...], strides:Tuple[int, ...]) -> List[Tuple[int, int]]:
+def to_shape_strides(shape:Tuple[int, ...], strides:Tuple[int, ...]) -> Tuple[Tuple[int, int], ...]:
   assert len(shape) == len(strides)
   ret = [(shape[0], strides[0])] if len(shape) > 0 else []
   for i in range(1, len(shape)):
     if (strides[i] != 0 and ret[-1][1] == shape[i]*strides[i]) or ret[-1][0] == 1 or (strides[i] == 0 and ret[-1][1] == 0):
       ret[-1] = (ret[-1][0] * shape[i], strides[i])
     else:
       ret.append((shape[i], strides[i]))
-  return ret
+  return tuple(ret)
 
 @functools.lru_cache(maxsize=None)
-def is_contiguous(shape:Tuple[int, ...], strides:Tuple[int, ...]) -> bool: return all([s1 == s2 or s == 1 for s,s1,s2 in zip(shape, strides, strides_for_shape(shape))])
+def is_contiguous(shape:Tuple[int, ...], strides:Tuple[int, ...]) -> bool:   
+  for s,s1,s2 in zip(shape, strides, strides_for_shape(shape)):
+    if not (s == 1 or s1 == s2): return False
+  return True
 
 @functools.lru_cache(maxsize=None)
 def filter_strides(shape:Tuple[int, ...], strides:Tuple[int, ...]) -> Tuple[int, ...]:
-  new_strides = []
-  for stride, shp in zip(strides, shape):
-    if shp != 1: new_strides.append(stride)
-    else: new_strides.append(0)
-  return tuple(new_strides)
-
-class View:
-  __slots__ = "shape", "strides", "offset", "mask", "shape_strides", "contiguous"
-  def __init__(self, shape:Tuple[int, ...], strides:Tuple[int, ...], offset:int=0, mask:Optional[Tuple[Tuple[int, int], ...]]=None):
-    self.shape, self.offset = shape, offset
-    self.strides = filter_strides(shape, strides)
-    self.mask = mask
-    self.shape_strides = to_shape_strides(shape, self.strides)
-    self.contiguous: bool = offset == 0 and is_contiguous(shape, self.strides) and mask is None
-
-  def __repr__(self): return f"View({self.shape}, {self.strides}, {self.offset}, {self.mask})"
-  def key(self): return (self.shape, self.strides, self.offset, self.mask)
-
-  def expr_node_mask(self, idx, valid=None) -> Node:
+  return tuple([0 if shp == 1 else stride for stride, shp in zip(strides, shape)])
+
+class View(NamedTuple):
+  shape:Tuple[int, ...]
+  strides:Tuple[int, ...]
+  shape_strides:Tuple[Tuple[int, int], ...]
+  mask:Optional[Tuple[Tuple[int, int]]]=None
+  contiguous:bool=False
+  offset:int=0
+
+  @staticmethod
+  @functools.lru_cache(maxsize=None)
+  def create(shape, strides=None, offset=0, mask=None):
```

Comment:
This can't be the constructor?

Reply:
Not if we want a `View` to be a `NamedTuple`. We want to do things in the init, but cannot overwrite the init of a `NamedTuple`. A `NamedTuple` gives us a lot of fast niceties for free, such as hashing, comparison and __repr__ (all written in C, so everything is just faster). Come to think of it, all nodes should probably also become namedtuples where possible...

I have now changed it to an explicit non-class factory method, `create_view`.

Reply:
hmm, you really can't use `__init__` in a superclass of named tuple?

Reply:
You can wrap it with another class. Updated.